### PR TITLE
Support Rust TransferManager Upload File

### PR DESCRIPTION
**Pull request by waahm7**
*Description of changes:*
- Sync to the latest commit of the Rust transfer manager
- Add support for benchmarking uploads

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: runners/s3-benchrunner-rust/src/transfer_manager.rs [hunk 1]
```diff
@@ -113,6 +139,31 @@ impl TransferManagerRunner {
 
         Ok(())
     }
+
+    async fn upload(&self, task_config: &TaskConfig) -> Result<()> {
+        let key = &task_config.key;
+
+        let stream = if self.config().workload.files_on_disk {
+            InputStream::from_path(key).with_context(|| "Failed to create stream")?
```
Comment (aajtodd) [inline runners/s3-benchrunner-rust/src/transfer_manager.rs:147]: Is the `key` actually the path here?

**Review by aajtodd — APPROVED** [review]
Looks reasonable, nice work.

File: runners/s3-benchrunner-rust/src/transfer_manager.rs (hunk 1 above)
Comment (waahm7) [reply runners/s3-benchrunner-rust/src/transfer_manager.rs:147]: Yes, key and path are supposed to be the same. https://github.com/awslabs/aws-crt-s3-benchmarks/blob/main/workloads/download-30GiB-1x.run.json#L11

File: runners/s3-benchrunner-rust/src/transfer_manager.rs [hunk 2]
```diff
@@ -1,20 +1,20 @@
...
 };
-use aws_sdk_s3::operation::get_object::builders::GetObjectInputBuilder;
 use tokio::fs::File;
 use tokio::io::AsyncWriteExt;
 use tokio::task::JoinSet;
 
 use crate::{
     BenchmarkConfig, Result, RunBenchmark, RunnerError, TaskAction, TaskConfig, PART_SIZE,
 };
+use bytes::Bytes;
```
Comment (graebm) [inline runners/s3-benchrunner-rust/src/transfer_manager.rs:17]: utterly trivial: move this up with the other 3rdparty uses

File: runners/s3-benchrunner-rust/src/transfer_manager.rs [hunk 3]
```diff
@@ -39,47 +40,72 @@ impl TransferManagerRunner {
...
+                .filter(|task| task.action == TaskAction::Upload)
+                .map(|task| task.size)
+                .max()
+                .unwrap_or(0)
+                .try_into()
+                .unwrap()
+        };
+        let random_data_for_upload: Bytes = {
+            let mut rng = fastrand::Rng::new();
+            let data: Vec<u8> = repeat_with(|| rng.u8(..)).take(upload_data_size).collect();
```
Comment (graebm) [inline runners/s3-benchrunner-rust/src/transfer_manager.rs:60]: this all works OK if upload_data_size is 0?

File: runners/s3-benchrunner-rust/src/transfer_manager.rs [hunk 4]
```diff
@@ -39,47 +40,72 @@ impl TransferManagerRunner {
...
+            data.into()
```
Comment (graebm) [inline runners/s3-benchrunner-rust/src/transfer_manager.rs:61]: trivial: I know in the C-runner, it's a big noticeable pause when filling a 100GiB buffer with random data. Would this go faster if we asked for random 64bit numbers? Or used a different library?

File: runners/s3-benchrunner-rust/src/transfer_manager.rs [hunk 5]
```diff
@@ -39,47 +40,72 @@ impl TransferManagerRunner {
...
-            .downloader
-            .download(input.into())
+            .transfer_manager
+            .download()
+            .bucket(&self.config().bucket)
+            .key(key)
+            .send()
             .await
-            .with_context(|| format!("failed starting download: {key}"))?;
+            .with_context(|| format!("failed downloading: {key}"))?;
```
Comment (graebm) [inline runners/s3-benchrunner-rust/src/transfer_manager.rs:108]: trivial: the previous text was more clear where in the process that download failed

**Review by graebm — APPROVED** [review]

File: runners/s3-benchrunner-rust/src/transfer_manager.rs (hunk 3 above)
Comment (waahm7) [reply runners/s3-benchrunner-rust/src/transfer_manager.rs:60]: I did a test with files_on_disk, which resulted in a 0-byte buffer, and it worked fine.

File: runners/s3-benchrunner-rust/src/transfer_manager.rs (hunk 5 above)
Comment (waahm7) [reply runners/s3-benchrunner-rust/src/transfer_manager.rs]: Thanks, reverted.

File: runners/s3-benchrunner-rust/src/transfer_manager.rs (hunk 4 above)
Comment (waahm7) [reply runners/s3-benchrunner-rust/src/transfer_manager.rs:64]: image

**Review by graebm — APPROVED** [review]

**Pull request merged**